appengine-deploy/src/main.ts

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as core from '@actions/core';
+import * as exec from '@actions/exec';
+import * as setupGcloud from '../../pkg/dist/index';
+
+async function run(): Promise<void> {
+  try {
+    // Get action inputs.
+    const projectId = core.getInput('project_id', { required: true });
+    const deliverables = core.getInput('deliverables');
+    const imageUrl = core.getInput('image-url');
+    const version = core.getInput('version');
+    const promote = core.getInput('promote');
+
+    if (!setupGcloud.isInstalled()) {
```

Comment:
My initial thought is two fold:

1. If we're going to use gcloud, we should clearly document the requirement on the `setup-gcloud` action and throw an error here if it's not installed instead of duplicating the logic.

2. We shouldn't depend on gcloud here unless we absolutely have to. It looks like we can create an AppEngine deployment using the API directly: https://cloud.google.com/appengine/docs/admin-api/deploying-overview

Reply:
1. There isn't a dependency on `setup-gcloud`. It can be used with or without, so I check if gcloud is installed. 
2. I debated `gcloud` vs using the API directly. Using the API directly means reimplementing gcloud functionality, such as: Uploading files to GCS, Authentication, yaml parsing, Creating and managing the GAE application services and versions, Java specifics (deploying appengine-web.xml, jars), and gcloud flags (promote/no promote, flex container images, version naming, multiple deployables).

Reply:
Sure, but you're doing a _lot_ of duplication of setup-gcloud here, including downloading and installing. It seems like we should remove all of that and require the user has the `setup-gcloud` action in their pipeline. We can just fail if you don't find the `gcloud` CLI command.

Reply:
Right, I've created a separate package to host the shared code between setup-gcloud and this action. After I was going to refactor the setup-gcloud method. I didn't want the action to require the setup-gcloud action. I am definitely open to switching it over if you think that is the best.

Reply:
It's my _opinion_ that setup-gcloud should be the only thing that installs/configures gcloud. If other actions require the `gcloud` CLI or authentication, they should document that requirement, but not duplicate the installation. I think it would be surprising to users to suddenly see a new tool in their path.

This also matches the way other actions work. For example, if your action requires python, you need to first use the `setup-python` action. Those actions don't magically install python because that would be unexpected.